File: FFT/Lab2/CUDA/c/cufftCUDA.c
#include "../../Inputs/c/fileReader.h"
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
// CUDA includes
#include <cufft.h>
#include <cuda.h>
#include <cuda_runtime.h>

// Input Size
#define SIZE 134217728

typedef double2 Complex; // CUFFT special type for complex data

//Device allocated pointers
Complex *d_inputFFT, *d_outputFFT, *d_outputInvFFT;

//Host allocated pointers for printing
Complex *h_inputFFT, *h_outputInvFFT;

//Timing Variables
cudaEvent_t start, stop;
float execTime = 0.0;


void allocate() {

    cudaEventCreate(&start);
    cudaEventCreate(&stop);

    cudaMalloc((void **)&d_inputFFT, sizeof(Complex) * SIZE);
    cudaMalloc((void **)&d_outputFFT, sizeof(Complex) * SIZE);
    cudaMalloc((void **)&d_outputInvFFT, sizeof(Complex) * SIZE);

    h_inputFFT = (Complex *) malloc(sizeof(Complex) * SIZE);
    h_outputInvFFT = (Complex *) malloc(sizeof(Complex) * SIZE);
}


void initialize() {

    int i;
    double inputs[SIZE];

    //Loading real numbers from a fle
    char *fname = "../../Inputs/complexInputs.txt";
    readInputs(fname, inputs, SIZE);

    for (i=0; i<SIZE; ++i) {
        h_inputFFT[i].x = inputs[i];
        h_inputFFT[i].y = 0.0;
    }

    cudaMemcpy(d_inputFFT, h_inputFFT, sizeof(Complex) * SIZE, cudaMemcpyHostToDevice);
}

void executeFFT() {

    int i;
    double error = 0.0;
    //Creating a plan for CUFFT
    cufftHandle plan;
    
    cudaEventRecord(start, 0);

    cufftPlan1d(&plan, SIZE, CUFFT_Z2Z, 1); // Z2Z is C2C with Double Precision

    //Performing FFT FORWARD
    cufftExecZ2Z(plan, (cufftDoubleComplex *) d_inputFFT,
                    (cufftDoubleComplex *) d_outputFFT,
                    CUFFT_FORWARD);

    //Performing FFT INVERSE
    cufftExecZ2Z(plan, (cufftDoubleComplex *) d_outputFFT,
        (cufftDoubleComplex *) d_outputInvFFT,
        CUFFT_INVERSE);

    cufftDestroy(plan);

    //Transfering data back to the CPU for comparison
    cudaMemcpy(h_outputInvFFT, d_outputInvFFT, sizeof(Complex) * SIZE, cudaMemcpyDeviceToHost);

    cudaEventRecord(stop, 0);
    cudaEventSynchronize(stop);

    //Accumulating total error
    for (i=0; i<SIZE; ++i) {
        error += fabs(h_inputFFT[i].x - h_outputInvFFT[i].x/SIZE);
    }

    printf("\nMean Error = %e\n", error / SIZE);

    cudaEventElapsedTime(&execTime, start, stop);
}

void deallocate() {

    cudaFree(d_inputFFT);
    cudaFree(d_outputFFT);
    cudaFree(d_outputInvFFT);
    
    free(h_inputFFT);
    free(h_outputInvFFT);
    
    cudaEventDestroy(start);
    cudaEventDestroy(stop);
}

int main() {

    allocate();
    
    initialize();
    
    executeFFT();
    
    printf("\nOnly FFT:");
    printf("\nExecution Time =  %.10f s\n", execTime / 1000);
    
    deallocate();
    
    return 0;
}

File: FFT/Lab2/CUDA/fortran/cufftCUDA.f90
program MAIN

    use fileReaderF90
    use cudafor
    use cufft
    
    implicit none
    integer, parameter :: SIGNAL_SIZE = 134217728
    double complex, dimension(:), pointer, device :: d_inputFFT, d_outputFFT, d_outputInvFFT
    double complex, dimension(:), pointer :: h_inputFFT, h_outputInvFFT
    
    type(cudaEvent) :: start, stop
    real :: execTime
    integer :: err

    !Creating events
    err = cudaEventCreate(start)
    err = cudaEventCreate(stop)

    allocate(h_inputFFT(SIGNAL_SIZE), h_outputInvFFT(SIGNAL_SIZE),&
    d_inputFFT(SIGNAL_SIZE), d_outputFFT(SIGNAL_SIZE), d_outputInvFFT(SIGNAL_SIZE))
    
    call initialize()
    
    call executeFFT()
    
    print *,"Only FFT:"
    print *, "Execution Time = ", execTime

    deallocate(h_inputFFT, h_outputInvFFT, &
    d_inputFFT, d_outputFFT, d_outputInvFFT)

    err = cudaEventDestroy(start)
    err = cudaEventDestroy(stop)

    contains

        subroutine initialize()
            implicit none
            
            integer :: i
            real*8, dimension(SIGNAL_SIZE) :: inputs
            
            ! Loading real numbers from a fle
            call readInputs("../../Inputs/complexInputs.txt", inputs, SIGNAL_SIZE);
            
            do i=1, SIGNAL_SIZE
                h_inputFFT(i) = dcmplx(inputs(i), 0.0)
            end do
            
            !cudaMemcpy
            d_inputFFT = h_inputFFT
        end subroutine initialize

        subroutine executeFFT()
            implicit none
            
            integer :: i, plan, err
            real*8 :: errorSum
            errorSum = 0.0
            
            err = cudaEventRecord(start, 0)
            !Create a cufft plan for Z2Z i.e complex to complex in Double precision
            err = cufftPlan1d(plan, SIGNAL_SIZE, CUFFT_Z2Z, 1)
            
            !Perform FFT out of place transform
            err = err + cufftExecZ2Z(plan, d_inputFFT, d_outputFFT, CUFFT_FORWARD)
            
            !Perform Inverse FFT in plance transform
            err = err + cufftExecZ2Z(plan, d_outputFFT, d_outputInvFFT, CUFFT_INVERSE)
            
            err = err + cufftDestroy(plan)
            
            !Cuda memcpy
            h_outputInvFFT = d_outputInvFFT
            
            err = cudaEventRecord(stop, 0)
            err = cudaEventSynchronize(stop)

            !Accumulate total error
            do i=1, SIGNAL_SIZE
                errorSum  = errorSum + DABS(real(h_inputFFT(i)) - real(h_outputInvFFT(i))/SIGNAL_SIZE)
            end do

            print *, "Mean Error: ", errorSum / SIGNAL_SIZE
            print *
            
            err = cudaEventElapsedTime(execTime, start, stop)
            execTime = execTime / 1000
        end subroutine executeFFT
    
end program MAIN
